temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java

```diff
@@ -290,17 +314,23 @@ public void close() {
       lock.unlock();
       return;
     }
-    // Do not close while runUntilAllBlocked executes.
-    // closeRequested tells it to call close() at the end.
     closeRequested = true;
-    if (inRunUntilAllBlocked) {
+    if (
+    // Do not close immediately if runUntilAllBlocked executes.
+    // closeRequested will tell a control thread to call close() at the end of the event loop.
+    inRunUntilAllBlocked
+        // some other thread or caller is already in the process of closing
+        || closeStarted) {
```

Comment:
> This will only be true if another thread gets to line 360 where this is unlocked right?
Yes, another thread started a shutdown, set the flag and released the lock, another thread can proceed to here and observe that another shutdown is already in progress.